### DISCUSSION TITLE
How to implement an email confirmation form field

### DISCUSSION THREAD

**Discussion by amerfarooq** [mantinedev/mantine#4528] [Q&A]
I have an array of borrowers in my form and I'm trying to figure out if it is possible to configure a validation rule wherein I can check if the `email` and `confirmEmail` match for EACH borrower.

```ts
const form = useForm({
    initialValues: {
      borrowers: [
        {
          firstName: '',
          lastName: '',
          email: '',
          confirmEmail: '',
          phoneNumber: {
            type: 'Cell',
            value: '',
            extension: '',
          },
          key: randomId(),
        },
      ],
    },
    validate: {
      borrowers: {
          email: isEmail('Email is invalid'),
          confirmEmail: isEmail('Email invalid'),
        },
      },
    },
  });
  ```

**Comment by amerfarooq** [comment]
One way I've thought up is as following:

```ts
  const form = useForm({
    initialValues: {
      borrowers: [
        {
          firstName: '',
          lastName: '',
          email: '',
          confirmEmail: '',
          phoneNumber: {
            type: 'Cell',
            value: '',
            extension: '',
          },
          key: randomId(),
        },
      ],
    },
    validate: {
      borrowers: {
        confirmEmail: (confirmEmail, formValues, path) => {
          const isEmailNotValid = isEmail()(confirmEmail);
          if (isEmailNotValid) {
            return 'Invalid Email';
          }
          const currentBorrowerIdx = parseInt(path.split('.')[1]);
          if (formValues.borrowers[currentBorrowerIdx].email !== confirmEmail) {
            return 'Confirm email does not match the provided email';
          }
        },
      },
    },
  });
  ```
  
  This does seem to work but is there a better way to do this?

**Comment by rtivital** [comment]
It is described in the documentation – https://mantine.dev/form/validation/#validation-based-on-other-form-values

**Comment by BenSilver1337** [comment]
To create a custom validation rule in your form that checks if the email and confirmEmail match for each borrower, you can use the validate function provided by useForm. In this function, you can iterate through the borrowers array and compare the email and confirmEmail fields. If they don't match, you can return an error message.

Here's how you can implement the custom validation:

jsx

import { useForm } from 'react-hook-form';

const form = useForm({
  initialValues: {
    borrowers: [
      {
        firstName: '',
        lastName: '',
        email: '',
        confirmEmail: '',
        phoneNumber: {
          type: 'Cell',
          value: '',
          extension: '',
        },
        key: randomId(),
      },
    ],
  },
  validate: (values) => {
    const errors = {};

    // Validate each borrower's email and confirmEmail
    values.borrowers.forEach((borrower, index) => {
      const { email, confirmEmail } = borrower;
      if (email !== confirmEmail) {
        // Set an error message for the confirmEmail field
        if (!errors.borrowers) {
          errors.borrowers = [];
        }
        if (!errors.borrowers[index]) {
          errors.borrowers[index] = {};
        }
        errors.borrowers[index].confirmEmail = 'Emails do not match';
      }
    });

    return errors;
  },
});

export default YourComponent;

In the validate function, we iterate through each borrower in the values.borrowers array. If the email and confirmEmail do not match for a borrower, we add an error message to the errors object, specific to that borrower's confirmEmail field.

With this setup, the form will show the error message "Emails do not match" for each borrower when the email and confirmEmail fields don't match. Remember to adjust the error messages and validation logic to fit your specific requirements.
